File: runme/mcgen/local/datan/solsrc/s1sd.c
#include <stdio.h>
#include <math.h>
#include <string.h>
#include "datsrc.h"
#include "grsrc.h"

/* Table of constant values */

static integer c__1 = 1, c__2 = 2, c__4 = 4, c__6 = 6, c__7 = 7,
               c__12 = 12, c__28 = 28;

main(void)
{
    /* Local variables */
    static integer iswit1, iswit, k, lcapt, n, kk, nn, nx, kmax, nws;
    static double alambd, delx, hist[1000], f, p, x0;
    static char capt[75], tx[75], ty[75];
    /* System generated locals */
    integer i__1;
    double d__1;
/* identify program to user */
    printf("%s\n", "Program S1SD provides interactively graphs of");
    printf("%s\n\n", "statistical functions of a discrete variable.");
/* ask for number of workstation */
    grnbws_();
    printf("%s\n", "Please, enter number of workstation:");
    printf("%s", "> ");
    scanf("%li", &nws);
    printf("%s\n", "Choose:");
    printf("%s\n", "1 - binomial distribution");
    printf("%s\n", "2 - hypergeometric distribution");
    printf("%s\n", "3 - Poisson distribution");
    printf("%s", "> ");
    scanf("%li", &iswit);
    printf("%s\n", "Choose:");
    printf("%s\n", "1 - probability");
    printf("%s\n", "2 - distribution function (cumulative prob.)");
    printf("%s", "> ");
    scanf("%li", &iswit1);
    memset(tx, ' ', 75);
    memset(ty, ' ', 75);
    memset(capt, ' ', 75);
    if (iswit == 1) {
/* binomial */
        printf("%s\n", "Enter p (0. <= p <= 1.):");
        printf("%s", "> ");
        scanf("%lf", &p);
        printf("%s\n", "Enter n (0 <= n <= 998):");
        printf("%s", "> ");
        scanf("%li", &n);
        (void)getchar();
        delx = 1.;
        if (iswit1 == 1) {
            strncpy(tx, "k", 1);
            strncpy(ty, "P(k)", 4);
            x0 = -.5;
            nx = n + 1;
        } else if (iswit1 == 2) {
            strncpy(tx, "K", 1);
            strncpy(ty, "P(k<K)", 6);
            x0 = -1.;
            nx = n + 2;
        }
        smhsin_(hist, &x0, &delx, &nx);
        sprintf(capt, "n = %4li", n);
        i__1 = n;
        for (k = 0; k <= i__1 + 1; ++k) {
            if (iswit1 == 1) {
                hist[k] = sdbinm_(&k, &n, &p);
            } else if (iswit1 == 2) {
                hist[k] = scbinm_(&k, &n, &p);
            }
        }
        lcapt = strlen(capt);
        smhsgr_(hist, &x0, &delx, &nx, tx, &c__1, ty, &c__6, capt, &lcapt,&nws);
     } else if (iswit == 2) {
/* hypergeometric */
        printf("%s\n", "Enter K (0 <= K <= 998):");
        printf("%s", "> ");
        scanf("%li", &kk);
        printf("%s\n", "Enter N (>= K):");
        printf("%s", "> ");
        scanf("%li", &nn);
        printf("%s\n", "Enter n (0 <= n <= N):");
        printf("%s", "> ");
        scanf("%li", &n);
        (void)getchar();
        kmax = min(n,kk);
        delx = 1.;
        nx = kmax + 2;
        sprintf(capt, "K = %4li, N = %4li, n = %4li", kk, nn, n);
        if (iswit1 == 1) {
            strncpy(tx, "k", 1);
            strncpy(ty, "P(k)", 4);
            x0 = -.5;
        } else if (iswit1 == 2) {
            strncpy(tx, "k'", 2);
            strncpy(ty, "P(k<k')", 7);
            x0 = -1.;
        }
        smhsin_(hist, &x0, &delx, &nx);
        i__1 = kmax;
        for (k = 0; k <= i__1 + 1;  ++k) {
            if (iswit1 == 1) {
                hist[k] = sdhypg_(&k, &n, &kk, &nn);
            } else if (iswit1 == 2) {
                hist[k] = schypg_(&k, &n, &kk, &nn);
            }
        }
        lcapt = strlen(capt);
        smhsgr_(hist, &x0, &delx, &nx, tx, &c__2, ty, &c__7, capt, &lcapt, &nws);
     } else if (iswit == 3) {
/* Poisson */
        printf("%s\n", "Enter lambda (0. < lambda < 10):");
        printf("%s", "> ");
        scanf("%lf", &alambd);
        (void)getchar();
        f = sdpois_(&k, &alambd);
        d__1 = alambd * 10.;
        kmax = nint(d__1);
        delx = 1.;
        nx = kmax + 1;
        sprintf(capt, "&l@ = %6.2lf", alambd);
        if (iswit1 == 1) {
            strncpy(tx, "k", 1);
            strncpy(ty, "P(k)", 4);
            x0 = -.5;
        } else if (iswit1 == 2) {
            strncpy(tx, "K", 1);
            strncpy(ty, "P(k<K)", 6);
            x0 = -1.;
        }
        smhsin_(hist, &x0, &delx, &nx);
        i__1 = kmax;
        for (k = 0; k <= i__1; ++k) {
            if (iswit1 == 1) {
                hist[k] = sdpois_(&k, &alambd);
            } else if (iswit1 == 2) {
                hist[k] = scpois_(&k, &alambd);
            }
        }
        lcapt=strlen(capt);
        smhsgr_(hist, &x0, &delx, &nx, tx, &c__1, ty, &c__6, capt, &lcapt,&nws);
    }
    return 0;
} /* main */

File: runme/mcgen/local/datan/solsrc/s1sd.f
      PROGRAM S1SD
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      DIMENSION HIST(1000)
      CHARACTER*75 TX,TY,CAPT
C identify program to user
      WRITE(*,*)' Program S1SD provides interactively graphs of'
      WRITE(*,*)' statistical functions of a discrete variable'
      WRITE(*,*)' '
C ask for number of workstation
      CALL GRNBWS()
      WRITE(*,*)' Please, enter number of workstation:'
      WRITE(*,*)' >'
      READ(*,*)NWS
      WRITE(*,*)' Choose'
      WRITE(*,*)' 1 - binomial distribution'
      WRITE(*,*)' 2 - hypergeometric distribution'
      WRITE(*,*)' 3 - Poisson distribution'
      WRITE(*,*)' >'
      READ(*,*)ISWIT
      WRITE(*,*)' Choose'
      WRITE(*,*)' 1 - probability'
      WRITE(*,*)' 2 - distribution function (cumulative prob.)'
      WRITE(*,*)' >'
      READ(*,*)ISWIT1
      IF(ISWIT.EQ.1) THEN
C binomial
        WRITE(*,*)' Enter p (0. <= p <= 1.)'
        WRITE(*,*)' >'
        READ(*,*)P
        WRITE(*,*)' Enter n (0 <=n <= 998)'
        WRITE(*,*)' >'
        READ(*,*)N
        DELX=1.D0
          CAPT='n = '
        IF(ISWIT1.EQ.1) THEN
          TX='k'
          TY='P(k)'
          X0=-0.5D0
          NX=N+1
        ELSE IF(ISWIT1.EQ.2) THEN
          TX='K'
          TY='P(k<K)'
          X0=-1.D0
          NX=N+2
        END IF
        CALL SMHSIN(HIST,X0,DELX,NX)
        WRITE(CAPT(5:8),'(I4)')N
        DO 10 K=0,N+1
          IF(ISWIT1.EQ.1) THEN
            HIST(K+1)=SDBINM(K,N,P)
          ELSE IF(ISWIT1.EQ.2) THEN
            HIST(K+1)=SCBINM(K,N,P)
          END IF
10      CONTINUE
        CALL SMHSGR(HIST,X0,DELX,NX,TX,1,TY,5,CAPT,8,NWS)
      ELSE IF(ISWIT.EQ.2) THEN
C hypergeometric
        WRITE(*,*)' Enter K (0 <= K <= 998)'
        WRITE(*,*)' >'
        READ(*,*)KK
        WRITE(*,*)' Enter N (>= K)'
        WRITE(*,*)' >'
        READ(*,*)NN
        WRITE(*,*)' Enter n (0 <= n <= N)'
        WRITE(*,*)' >'
        READ(*,*)N
        KMAX=MIN(N,KK)
        DELX=1.D0
        NX=KMAX+2
        CAPT='K = ****, N = ****, n = ****'
        WRITE(CAPT(5:8),'(I4)')KK
        WRITE(CAPT(15:18),'(I4)')NN
        WRITE(CAPT(25:28),'(I4)')N
        IF(ISWIT1.EQ.1) THEN
          TX='k'
          TY='P(k)'
          X0=-0.5D0
        ELSE IF(ISWIT1.EQ.2) THEN
          TX='k'''
          TY='P(k<k'')'
          X0=-1.D0
        END IF
        CALL SMHSIN(HIST,X0,DELX,NX)
        DO 20 K=0,KMAX+1
          IF(ISWIT1.EQ.1) THEN
            HIST(K+1)=SDHYPG(K,N,KK,NN)
          ELSE IF(ISWIT1.EQ.2) THEN
            HIST(K+1)=SCHYPG(K,N,KK,NN)
          END IF
20      CONTINUE
        CALL SMHSGR(HIST,X0,DELX,NX,TX,2,TY,8,CAPT,28,NWS)
      ELSE IF(ISWIT.EQ.3) THEN
C Poisson
        WRITE(*,*)' Enter lambda (0. < lambda < 10)'
        WRITE(*,*)' >'
        READ(*,*)ALAMBD
        F=SDPOIS(K,ALAMBD)
        KMAX=NINT(10.D0*ALAMBD)
        DELX=1.D0
        NX=KMAX+1
        CAPT='&l@ = ******'
        WRITE(CAPT(7:12),'(F6.2)')ALAMBD
        IF(ISWIT1.EQ.1) THEN
          TX='k'
          TY='P(k)'
          X0=-0.5D0
        ELSE IF(ISWIT1.EQ.2) THEN
          TX='K'
          TY='P(k<K)'
          X0=-1.D0
        END IF
        CALL SMHSIN(HIST,X0,DELX,NX)
        DO 30 K=0,KMAX
          IF(ISWIT1.EQ.1) THEN
            HIST(K+1)=SDPOIS(K,ALAMBD)
          ELSE IF(ISWIT1.EQ.2) THEN
            HIST(K+1)=SCPOIS(K,ALAMBD)
          END IF
30      CONTINUE
        CALL SMHSGR(HIST,X0,DELX,NX,TX,1,TY,6,CAPT,12,NWS)
      END IF
      END
